### PR TITLE
Julia Implementation

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

Aligning with SENA's mission, and one of the main purposes of this repository, this Draft PR will introduce an implementation of the `c_sw` kernel using The Julia Programming Language.  As this implementation is currently a work in progress, with more adjustments and refactoring anticipated in the near future, an official PR will be issued once the initial state of the Julia code is reviewed and improved upon. 

Updates to the Github Actions will build and test the Julia implementation. 

## Types of changes

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [x] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [x] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-c_sw/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

In this first Julia version, performance has not be thoroughly investigated. There will likely be additions and changes with regard to improving performance. 

Since there are many ways to structure Julia projects/repositories, a 'conventional' or most 'Julian' way of implementing `c_sw` is continuously being evaluated.

### CODE REVIEW

**Comment by christopherwharrop-noaa** [comment]
@robgonzalezpita - Can you please bring this PR up to date with develop and make sure it still works after doing so?

File: Julia/Manifest.toml [hunk 1]
```diff
@@ -0,0 +1,249 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (christopherwharrop-noaa) [inline Julia/Manifest.toml:1]: Is this file created automatically?  Is it something that should go into the repository?

File: Julia/test/data/inputs/inputs.toml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+[c_sw_48x48_3]
+  name = "c_sw_48x48_3"
+  input_data_dir = "test/data/inputs"
+  input_file = "../test/data/inputs/c_sw_48x48.nc"
+  output_data_dir = "Julia/test/data/outputs"
+  output_file = "test/data/outputs/c_sw_48x48_3.nc"
+  log_dir = "test_output"
+  test_out_file = "test/test_output/c_sw_48x48_3.test.out"
+  interpFactor = 3
```
Comment (christopherwharrop-noaa) [inline Julia/test/data/inputs/inputs.toml]: Need a newline here

File: Julia/test/unit_tests.jl [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        # create a new .test file, print some data to it, test that the data was writen correctly, delete the file 
+
+end
+
+
+@testset "All tests" begin
+    subroutine_tests()
+    # load_state_tests()
+    # write_print_tests()
+end
```
Comment (christopherwharrop-noaa) [inline Julia/test/unit_tests.jl]: Need a newline here

File: Julia/Manifest.toml (hunk 1 above)
Comment (robgpita) [reply Julia/Manifest.toml:1]: Yes, this file is created automatically, but it is standard to include it in Julia repositories in order to track the dependencies of the packages one uses in their project. The Manifest.toml and Project.toml together enable correct code loading via `Pkg.instantiate` . Here is some [further reading](https://pkgdocs.julialang.org/v1/toml-files/) if you are curious.

Comment (christopherwharrop-noaa) [reply Julia/Manifest.toml:1]: Thank you for the explanation and link.  Very helpful.

File: Julia/src/sw_core.jl [hunk 4]
```diff
@@ -0,0 +1,1009 @@
+#=
```
Comment (christopherwharrop-noaa) [inline Julia/src/sw_core.jl]: Would you mind adding some large-ish comment header blocks above each function to create an easily identifiable separator so readers can easily see where new function starts?  No need for super detailed descriptions in those header blocks, something like what was done for the Fortran version would be sufficient.

File: Julia/src/states.jl [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+      OffsetArray(Array(ds["wc"]),           isd:ied,      jsd:jed,     1:npz),
+      OffsetArray(Array(ds["ut"]),           isd:ied,      jsd:jed,     1:npz),
+      OffsetArray(Array(ds["vt"]),           isd:ied,      jsd:jed,     1:npz),
+      OffsetArray(Array(ds["divg_d"]),       isd:ied + 1,  jsd:jed + 1, 1:npz)
+    )
+
+    # Close the .nc file 
+    close(ds)
+
+  end # function State
```
Comment (christopherwharrop-noaa) [inline Julia/src/states.jl]: Need newline

**Comment by christopherwharrop-noaa** [comment]
@robgonzalezpita - This is looking really good!  I don't know enough about Julia to comment on technical merits of the Julia code itself.  There are a few syntax/style items, but I don't see any issues here.  I think this is just about ready to turn into a full PR.

File: Julia/src/sw_core.jl (hunk 4 above)
Comment (robgpita) [reply Julia/src/sw_core.jl]: Sure will! Skipped this for the majority of subroutines.

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
This is great work.  I was able to test this successfully on two of our supercomputers.  Initially I had a lot of trouble because I did not realize that when installing Julia, both the `lib/` and `lib/julia` directories needed to be on `LD_LIBRARY_PATH`.  Also, one complication is that to make this work on a system without an internet connection, one must first do the build on a system of the same OS/Architecture that does have access, and then copy over the resulting `~/.julia` directory to the system without internet access.  This will preload the dependencies so they don't need to be downloaded.

Could you please add some notes in the documentation about these two items?  Once that is done, I'll approve.

**Review by christopherwharrop-noaa — APPROVED** [review]
This is a nice accomplishment.  The first Julia implementation is ready to go in.

**Comment by robgpita** [comment]
@christopherwharrop-noaa Thanks for taking the time to get this running on the HPCs and describing how you overcame the encountered issues!

**Pull request merged**